src/Psl/Type/Internal/ShapeType.php

```diff
@@ -28,12 +23,15 @@ final class ShapeType extends Type\Type
      */
     private array $elements_types;
 
+    private bool $allow_unknown_fields;
+
     /**
      * @param array<Tk, Type\TypeInterface<Tv>> $elements_types
      */
-    public function __construct(array $elements_types)
+    public function __construct(array $elements_types, bool $allow_unknown_fields = false)
     {
         $this->elements_types = $elements_types;
+        $this->allow_unknown_fields = $allow_unknown_fields;
```

Comment:
What happens type-wise if you set this prop to true?
Should it be typed as the shape you provide or as the more generic default as templates above the class?

Reply:
currently nothing. this is more of a runtime configuration.

Reply:
Alright, so if you access an undefined key directly, psalm will lose track of the value type and you'll need to set it manually?

Reply:
this better explains it:
```php
$result = Json\typed($content, Type\shape([
  'username' => Type\non_empty_string(),
  'password' => Type\non_empty_string(),
  'country' => Type\optional(Type\non_empty_string())
], true));

// $result = array{username: non-empty-string, password: non-empty-string, country?: non-empty-string}

$username = $result['username']; // psalm: non-empty-string
$password  = $result['password']; // psalm: non-empty-string
$country = $result['country']; // psalm error: 'country' might not be defined.
$country  = $result['country'] ?? null; // psalm: non-empty-string|null
$bar = $result['foo']; // psalm error: 'foo' is not defined
$bar = $result['foo'] ?? null; // psalm error: unable to infer the type $bar is being assigned to.
/** @var string|null $bar */
$bar = $result['foo'] ?? null; // psalm: string|null
```

at the end, psalm will still remember that `$result` is `array{username: non-empty-string, password: non-empty-string, country?: non-empty-string}`, however, accessing fields that are not specified in the original type will result in an error, using `?? null` ( or `isset($shape['foo'])` ), psalm will tell you about another error ( unable to infer the type the variable is being assigned to ), so in that case, you have to either: 

a. use `@var`
b. use `Type\` function to assert it's type.

Reply:
Alright, exactly what I had in mind. makes sense! Thanks for the info